### ISSUE TITLE
Add `batch_size` parameter to `Task.map`

### ISSUE THREAD

**Issue by evamaxfield** [PrefectHQ/prefect#2459]
## Use Case
*Please provide a use case to help us understand your request in context*
When you have hundreds of thousands of tasks to map against, it may overload or take a very long time for the DaskExecutor (and the underlying Dask scheduler) to start up.

## Solution
*Please describe your ideal solution*
```python
results = my_task.map(*iterables, batch_size=10)
```

`batch_size` is `None` but can be set to _completely_ process the data in that batch prior to moving on to the next batch.

I emphasize this because [distributed just added a `batch_size` parameter](https://github.com/dask/distributed/pull/3650) that _adds_ tasks to the scheduler in batches rather than fully processes items in batches.

## Alternatives
*Please describe any alternatives you've considered, even if you've dismissed them*
Some context on why the "fully process this batch vs add this batch" is important. We use a library to read images that utilizes dask under the hood. In doing so, if we want to process thousands of images, each image in the `map` call may result in thousands more tasks. So even if the data is _added_ in batches to the scheduler, as those batches are added, they are creating thousands of tasks themselves, which overloads the scheduler anyway

**Comment by joshmeek** [comment]
Thanks for opening the feature request @JacksonMaxfield! I like the idea and I don't _think_ there would be anything preventing it since it all happens on the Dask side of things but I could be missing something. I am wondering how this `batch_size` kwarg would fair if not using Dask. i.e. mapping with using the LocalExecutor executes synchronously so the batch size would essentially be ignored.

cc @cicdw

**Comment by jcrist** [comment]
Hi @JacksonMaxfield. With the DFE mapping refactor, mapping no longer is a layer on top of dask's `map` (Prefect walks its own `dag`, submitting components to the executor (in this case dask) individually).

The `batch_size` argument to dask's `map` is useful for reducing the size of the submitted graphs, sending the graph in batches rather than all at once. Previously dask's `map` would submit all the mapped tasks at once, possibly submitting a very large graph. Since prefect now just uses `submit` for everything, the individual components will always be small - we have more `submit` calls, but we'll never submit a single large graph.

Also note that prefect tasks (at least when using cloud/server) have some level of overhead on top of them. If batching is intended to reduce the overhead of `executor` calls, I don't see this resulting in a perf improvement after the DFE refactor - the api calls to orchestration would likely dominate the faster submit to the executor.

Since supporting a `batch_size`-type-thing is tricky for us to do, and may have limited benefit, I'm tempted to close this. Thoughts @JacksonMaxfield?

**Comment by JacksonMaxfield** [comment]
Hey @jcrist thanks for the update. I've been following the Dask optimizations and such and love what I am seeing. I think in general the direct need for this is low for my day-to-day workloads so agree that it has limited benefits. It is also partially caused by our own image reading library so fault to ourselves.

That all said, I do think there is value in having something like this but I would understand closing the issue as it is low priority, would likely be low usage except in the few cases like myself, and as you said would take a decent amount of work. So no worries if you close. If I find I really need it in a couple months, I may take a stab at it myself and only then ask for help :joy:.

**Comment by gregjohnso** [comment]
+1 this would pretty value for when I want to map an iterable of 300k+

**Comment by joelluijmes** [comment]
I’m looking for something similar as asked here (batch mapped tasks or nested mapping). I'm wondering if there are any updates on this @jcrist? I think it would be very interesting feature for Prefect 🤓 

What I’m after is:
1. Retrieve some dynamic list (i.e. query from database)
2. Batch the result set in 20ish items to process in parallel
3. For each item in the batch, do a branch of chained tasks
4. Wait for batch to complete, repeat for next batch until exhausted

Basically I want to write something as

```python
tasks = dynamic_list_of_tasks()
windowed_tasks = fixed_window(tasks, window_size=5)

def process_item(item):
    x = task_1()
    task_2(x)

def process_window(tasks):
    apply_map(process_item, tasks)

apply_map(process_window, windowed_tasks)
```

Or something proposed above, that it automatically processes the list in batches.

Full disclosure, I already asked it in Slack but haven't got a reply yet. Additionally this has come up earlier:
* https://prefect-community.slack.com/archives/CL09KU1K7/p1613660367492400 (my question)
* https://prefect-community.slack.com/archives/C014Z8DPDSR/p1612382522176200
* https://prefect-community.slack.com/archives/CL09KU1K7/p1608233129249800

**Comment by ChappIO** [comment]
I have the same issue. I need to process tens of thousands of entries yet my api can only handle about 100 entries at a time. So I would like to:

1. Extract all entries
2. Map each entry to a transform task to gather additional information
3. Run my loading task for each 100 items in the iterator. So if I have 98 thousand items in the iterator I want to call "load" 980 times with 100 items.

**Comment by madkinsz** [comment]
@ChappIO 

Is a task-based fixed window sufficient for your case? 
```
import random

import prefect
from prefect import Flow, task, Parameter

MAX_SIZE = 30


@task
def dynamic_list_of_tasks(window_size):
    size = random.randint(window_size, MAX_SIZE)
    size += size % window_size  # Ensure it is divisable for the fixed window
    return [i for i in range(size)]


@task
def fixed_window(tasks: list, window_size: int):
    windowed = []
    while tasks:
        windowed.append(tasks[:window_size])
        tasks = tasks[window_size:]
    return windowed


@task
def display(i):
    prefect.context.logger.info(f"Processing {i}")


window_size = Parameter("window_size", default=2)

with Flow("windowed-tasks") as flow:
    tasks = dynamic_list_of_tasks(window_size)
    windowed_tasks = fixed_window(tasks, window_size)
    display.map(windowed_tasks)

flow.run()
```

The downside of an approach like this is that does not support DFE, but that does not matter for everyone.

**Comment by ChappIO** [comment]
I don't think this will do it for me because it requires me to have a huge amount of memory available to keep all the tasks in memory. I have an iterator (a mapped task) which I would like to pass to the next task in batches of a maximum size:

Example:

```
@task
def extract():
    return range(550000000)
    
@task
def transform(i): 
   return i * 10
   
@task
def load(batch)
   // Here I would like a batch of 100
   do_a_thing(batch)
   
with Flow("Example") as flow:
  items = extract()
  mapped = transform.map(i=items)
  load.map(batch=somehow_batch_task_items(mapped))
```
